src/api/ApiResponder.php

```diff
@@ -651,6 +610,72 @@ public function process_request($args) {
             echo json_encode($output);
         }
     }
+
+    // This function looks at the provided arguments and verifies
+    // both that an appropriate interface routine exists and that
+    // the requester has sufficient credentials to access it
+    protected function verify_function_access($args) {
+        if (array_key_exists('type', $args)) {
+            $funcname = 'get_interface_response_' . $args['type'];
+            if (method_exists($this, $funcname)) {
+                if (in_array($args['type'], $this->unauthFunctions)) {
+                    $result = array(
+                        'ok' => TRUE,
+                        'functype' => 'newuser',
+                        'funcname' => $funcname,
+                    );
+                } elseif (auth_session_exists()) {
+                    $result = array(
+                        'ok' => TRUE,
+                        'functype' => 'auth',
+                        'funcname' => $funcname,
+                    );
+                } else {
+                    $result = array(
+                        'ok' => FALSE,
+                        'message' => "You need to login before calling API function " . $args['type'],
+                    );
+                }
+            } else {
+                $result = array(
+                    'ok' => FALSE,
+                    'message' => 'Specified API function does not exist',
+                );
+            }
+        } else {
+            $result = array(
+                'ok' => FALSE,
+                'message' => 'No "type" argument specified',
+            );
+        }
+        return $result;
+    }
+
+    protected function create_interface($args, $check) {
+        if ($check['functype'] != 'auth') {
+            return new BMInterfaceNewuser($this->isTest);
+        }
+
+        apache_note('BMUserID', $_SESSION['user_id']);
+
+        $interface = $this->select_interface_type($check['funcname']);
+
+        if (!isset($args['automatedApiCall']) || $args['automatedApiCall'] != 'true') {
+            $interface->player()->update_last_access_time($_SESSION['user_id']);
+        }
+
+        return $interface;
+    }
+
+    protected function select_interface_type($funcname) {
+        if (array_search($funcname, $this->playerFunctions, TRUE)) {
+            $interface = new BMInterfacePlayer($this->isTest);
+        } else {
+            $interface = new BMInterface($this->isTest);
+        }
+
+        return $interface;
+    }
```

Comment:
Not totally sure this is necessary --- in cases like the createGame case, can you just call `$interface->player()->create_game()`?  Ah, looking at more of the code, i see that at least part of the problem is that, in the createGame case, too much logic is in ApiResponder.  I wonder if this would be a good time to move more logic to BMInterfacePlayer, but i also don't think it's strictly necessary.

Anyway.  I think this is fine, and it is approximately what i had in mind (to the extent that i can remember) --- it's just a little untidy that every function needs to call `$interface->player()` to set the modification time.  Makes me wonder if all authenticated functions should instantiate both `$this->interface` and $this->interfacePlayer` or something.